### PR TITLE
Fix opening projects

### PR DESCRIPTION
**Pull request by MrFlashAccount**
#### Tl;dr 
- Closes: enso-org/cloud-v2#1338 
This PR fixes bugs with opened projects. Now all projects close/open properly and list of opened projects stored in the single place

---

#### Context:
Few sentences on the high level context for the change. Link to relevant design docs or discussion.

#### This Change:
What this change does in the larger context. Specific details to highlight for review:
1. Removes a bunch of useEffects across the Dashboard page
2. Project status now a react-query state, can be reused across the app
3. Eliminated the need of `waitIntilProjectIsOpened`

---

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

File: app/ide-desktop/lib/dashboard/src/components/dashboard/Permission.tsx [hunk 1]
```diff
@@ -38,7 +38,10 @@ const ASSET_TYPE_TO_TEXT_ID: Readonly<Record<backendModule.AssetType, text.TextI
 /** Props for a {@link Permission}. */
 export interface PermissionProps {
   readonly backend: Backend
-  readonly asset: backendModule.Asset
+  readonly asset: {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/dashboard/Permission.tsx]: a bit awkward imo, i think it's better if it's flattened. like i'd understand if this was perhaps an existing type, but as it is you specify it as `asset={{ id, type }}` which just feels like a weird mix of jsx and regular js

File: app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx [hunk 2]
```diff
@@ -44,8 +44,11 @@ const TYPE_SELECTOR_Y_OFFSET_PX = 32
 export interface ManagePermissionsModalProps<
   Asset extends backendModule.AnyAsset = backendModule.AnyAsset,
 > {
-  readonly backend: Backend
-  readonly item: Asset
+  readonly item: {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx]: similarly here re: nesting. perhaps these should be `Pick<backendModule.AnyAsset, 'a' | 'b' | 'c'>`?

File: app/ide-desktop/lib/dashboard/src/text/english.json [hunk 3]
```diff
@@ -306,6 +306,7 @@
   "openInfoMenu": "Open info menu",
   "noProjectIsCurrentlyOpen": "No project is currently open.",
   "offlineTitle": "You are offline",
+  "offlineErrorMessage": "It seems like you're offline. Please make sure you're connected to the internet and try again",
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/text/english.json]: contractions like "you're" tend to sound informal, prefer the expanded form ("you are")

File: app/ide-desktop/lib/dashboard/src/utilities/newtype.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@
 // ===============
 
 /** An interface specifying the variant of a newtype. */
-interface NewtypeVariant<TypeName extends string> {
+export interface NewtypeVariant<TypeName extends string> {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/utilities/newtype.ts:8]: do these need to be `export`ed?

File: app/ide-desktop/lib/dashboard/src/layouts/AssetsTable.tsx [hunk 5]
```diff
@@ -2210,13 +2247,26 @@ export default function AssetsTable(props: AssetsTableProps) {
...
       const isSoleSelected = selectedKeys.size === 1 && isSelected
+
       return (
         <AssetRow
           key={key}
+          updateAssetRef={instance => {
+            if (instance != null) {
+              updateAssetRef.current[item.item.id] = instance
+            } else {
+              // this is hacky, but it works
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetsTable.tsx]: nit: full sentences (capital + period)

also, wow it really is hacky lmao

File: app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx [hunk 6]
```diff
@@ -193,21 +196,13 @@ export function Tab(props: InternalTabProps) {
...
-        },
-        () => {
-          setIsLoading(false)
-        }
-      )
-    } else {
-      setIsLoading(false)
-    }
-  }, [loadingPromise])
+  const { isLoading, data } = reactQuery.useQuery<backend.Project>(
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx:201]: i really don't know whether it makes sense for a `Tab` of all things to know about the existence of `project`s. is it perhaps possible to provide an optional `UseQueryOptions` instead?

File: app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectIcon.tsx [hunk 7]
```diff
@@ -65,226 +59,93 @@ const LOCAL_SPINNER_STATE: Readonly<Record<backendModule.ProjectState, spinner.S
...
-  const closeProjectMutation = backendHooks.useBackendMutation(backend, 'closeProject')
-  const getProjectDetailsMutation = backendHooks.useBackendMutation(backend, 'getProjectDetails')
-  const waitUntilProjectIsReadyMutation = backendHooks.useBackendMutation(
-    backend,
-    'waitUntilProjectIsReady'
-  )
-  const openProjectMutate = openProjectMutation.mutateAsync
-  const getProjectDetailsMutate = getProjectDetailsMutation.mutateAsync
+  const state = (() => {
+    // project is not opened
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectIcon.tsx]: nit: prefer full sentences

File: app/ide-desktop/lib/dashboard/src/utilities/newtype.ts (hunk 4 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/utilities/newtype.ts:8]: Yes, because otherwise ts can't access them and show an error https://stackoverflow.com/questions/43900035/ts4023-exported-variable-x-has-or-is-using-name-y-from-external-module-but

File: app/ide-desktop/lib/dashboard/src/layouts/Drive.tsx [hunk 8]
```diff
@@ -68,16 +72,29 @@ export interface DriveProps {
   readonly dispatchAssetListEvent: (directoryEvent: assetListEvent.AssetListEvent) => void
   readonly assetEvents: assetEvent.AssetEvent[]
   readonly dispatchAssetEvent: (directoryEvent: assetEvent.AssetEvent) => void
-  readonly setProjectStartupInfo: (projectStartupInfo: backendModule.ProjectStartupInfo) => void
-  readonly doOpenEditor: () => void
-  readonly doCloseEditor: (projectId: backendModule.ProjectId) => void
+  readonly doOpenEditor: (id: dashboard.Project['id']) => void
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/layouts/Drive.tsx]: prefer exporting `ProjectId` and using that instead just for consistency with `backend.ProjectId`

File: app/ide-desktop/lib/dashboard/src/utilities/newtype.ts (hunk 4 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/utilities/newtype.ts:8]: i guess this is one usecase for using `type`s :skull:

File: app/ide-desktop/lib/dashboard/src/pages/dashboard/Dashboard.tsx [hunk 9]
```diff
@@ -320,40 +361,50 @@ export default function Dashboard(props: DashboardProps) {
...
+
+  const doRemoveSelf = eventCallbacks.useEventCallback((project: Project) => {
+    dispatchAssetListEvent({ type: AssetListEventType.removeSelf, id: project.id })
+    doCloseProject(project)
+  })
+
+  const onSignOut = eventCallbacks.useEventCallback(() => {
+    setPage(TabType.drive)
+
+    launchedProjects.forEach(project => {
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/pages/dashboard/Dashboard.tsx]: prefer a `for...of...` loop, even if just for consistency

File: app/ide-desktop/lib/dashboard/src/pages/dashboard/Dashboard.tsx [hunk 10]
```diff
@@ -414,42 +463,81 @@ export default function Dashboard(props: DashboardProps) {
                 </tabBar.Tab>
               )}
             </TabBar>
+
             <UserBar
-              backend={
-                projectStartupInfo?.backendType === backendModule.BackendType.remote
-                  ? remoteBackend
-                  : localBackend
+              onShareClick={
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/pages/dashboard/Dashboard.tsx]: this is a pretty big callback, maybe it should be extracted out to a variable?

**Review by somebody1234 — APPROVED** [review]
CR approved, minor nits noted above

File: app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx (hunk 6 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx:201]: good idea, thanks for noticing!

File: app/ide-desktop/lib/dashboard/src/components/dashboard/Permission.tsx (hunk 1 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/dashboard/Permission.tsx]: > specify it as asset={{ id, type }} which just feels like a weird mix of jsx and regular js

Yeah, but I introduced this type to declare exactly which fields it expects as an input. We still can pass assets as is. IMO we'd use a format like that `asset={{ id, type }}` only inside tests, and it's good IMO because we don't need to maintain unneeded fields in mocks.

File: app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx]: Questionable, on the one hand - this makes sense, on the other hand, it leads to unnecessary coupling between a backend and the modal. I think it might be a good idea to decouple these things as much as we can to be able to reuse this modal in mocked environments and other places. WDYT?

File: app/ide-desktop/lib/dashboard/src/text/english.json (hunk 3 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/text/english.json]: Thanks!

File: app/ide-desktop/lib/dashboard/src/components/dashboard/Permission.tsx (hunk 1 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/components/dashboard/Permission.tsx]: see other comment... somewhere else then - perhaps `Pick` might be a better idea? for multiple reasons:
- to let the user know "this is a subset of `AnyAsset`, so you can pass an `AnyAsset` here
- so that it never goes out of sync with `AnyAsset`

File: app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx]: while that makes sense, i think it does make sense to couple them *somewhat*, at least - otherwise you can no longer pass `asset`s directly to these functions, right? like, *technically* there is a dependency here. but to the user, they can't see the keys that have been picked away so it should make no difference. `backendModule` is almost all just type declarations anyway - there is one class, but it's an abstract class so i think it basically counts as a type as well. similarly with the constructors for the newtypes.

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx]: tl;dr:
- as long as we `import type * as backendModule`, it shouldn't affect mocking without having to import the actual module at runtime
- we want it to be coupled so that we can pass `asset`s directly

worth noting that we *techincally* already do something similar when picking `pageX` and `pageY` or whatever from `MouseEvent` from certain fixed-positioned modals.

File: app/ide-desktop/lib/dashboard/src/pages/dashboard/Dashboard.tsx (hunk 9 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/pages/dashboard/Dashboard.tsx]: I'm not sure that just banning `foreach` in favor of `forof` is a good idea... If we don't have strong arguments against using a particular syntax, it's up to the author of the PR, right?

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/pages/dashboard/Dashboard.tsx]: On the second thought we can simplify this to just `doCloseAllProject`

File: app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx]: > while that makes sense, i think it does make sense to couple them somewhat, at least - otherwise you can no longer pass assets directly to these functions, right?

I guess not, due to the duck-typing system in ts we should be able to pass anything that extends the declared shape, no matter whether explicitly or implicitly.

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx]: So the real question is more like: do we _want_ to explicitly say that this modal should work only with assets, or with _anything_ that has these fields? Shall we use the same modal to assign permissions to assets, and also, maybe assign permissions to users to access certain pages, e.g. manage users/groups/subscriptions? 
It's a good question and I don't have an answer here

File: app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx (hunk 6 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx:201]: On the other hand, we can't decouple things here because right after we have:
```  typescript
const isFetching = isLoading || (data && data.state.type !== backend.ProjectState.opened)
```

Which couples things together. And I'm not sure how we can bypass this. maybe introduce another function that will only expose `isLoading`? or make a wrapper... Not sure 🤔

Ideally, if we want to make `tab` generic, we need to extract the `core` part into a separate component.

File: app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx]: >  we should be able to pass anything that extends the declared shape

exactly my point - i just mean that assets are the "first-class" users of this modal, right? i'm just thinking, if it goes out of sync with the assets then we'd have to do `{ foo2: asset.foo, bar: asset.bar }` or similar rather than the more intuitive `asset`?

or i guess iow: i think assets should be the "source of truth" of what the shape looks like - `Pick` omits all keys other than the relevant ones anyway, so as long as the keys are there the code will typecheck



but perhaps you mean it should be decoupled further? maybe we should be providing e.g. `setPermissions` as a prop? but i suppose it's probably not particularly relevant right now, and we can probably refactor later on when we're sure we need to use it for other things?

File: app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx (hunk 6 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx:201]: i mean, we could have an optional override prop `isLoading: (query: T) => boolean`? which is... a bit ugly perhaps, it does *work*...

a more correct (?) way would probably be to make a *different* query with a `queryFn` that only resolves when the project is opened (and rejects if the project state becomes one of the closed states). not sure how feasible that is though?

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/TabBar.tsx:201]: > different query with a queryFn that only resolves when the project is opened (and rejects if the project state becomes one of the closed states). not sure how feasible that is though?

Had a thought on this, but not sure how we cancel this query if we close the tab before the project is opened...
btw this is the exact reason why I don't use `waitForProjectOpen()` on the remote backend

File: app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/ManagePermissionsModal.tsx]: Lets use pick then :)

**Comment by PabloBuchu** [comment]
The original problem is gone. Thank you! But it looks like there is still problem with local project state refresh (to me it feels the local projects take much longer time to open:
1. When I interrupt opening local projects with opening another one it won't change it state to ready to open. I need to manually refresh the state by changing the page
2. Having an opened project and switching back and forth with another one messes state a bit (see video at the beginning)

https://github.com/enso-org/enso/assets/2855109/cbc919cf-6c10-45d6-96f5-85302e67d651

**Comment by MrFlashAccount** [comment]
@PabloBuchu thank you! I've addressed the issues you've raised

**Review by PabloBuchu — APPROVED** [review]

**Comment by PabloBuchu** [comment]
QA ✅
